### PR TITLE
Devnet contracts

### LINKED ISSUE #381: Add instantiation of required contracts to sunodo/devnet
## 📚 Context

Rollups 1.0 used to have an Authority and a History contract already deployed and included in the npm package.
Those contracts are necessary to run a node.

Rollups 1.2 does not include those contract any longer. It's a responsibility of the node runner to deploy those contracts prior to running a node.

The devnet environment could do that for the user, and have a docker image with anvil ready to be used by a local development environment.

The upcoming marketplace feature will also need a ValidatorNodeProvider contract instantiated.

## ✔️ Solution

Write a hardhat task in the devnet project that instantiates the required contracts.
Call that task during build to produce the anvil dump we are baking into a docker anvil image.
We also need to store the address of the contracts in some way inside the image.

## 📈 Subtasks

- [x] #384
- [x] write a task to instantiate a ValidatorNodeProvider
- [x] modify build process to call the tasks

## 🎯 Definition of Done

- [x] sunodo/devnet image has information about instantiated contracts

### PR DESCRIPTION
**Pull request by tuler**
This creates a `Devnet` contract to the `devnet` project that instantiates an authority, a history and a provider, with simple defaults for a devnet environment.

The `pnpm run build --filter devnet` from root does the following:

1. build `contracts` project and `token` project, as they are dependencies to the `devnet` project.
2. compile solidity contract of devnet project
3. run an anvil instance with --dump-state (on exit)
4. deploy contracts to anvil through hardhat-deploy external script (cartesi contracts, sunodo contracts, sunodo token), exporting those to a json
5. deploy an instance of the `Devnet` contract
6. read addresses of authority, history, provider from `Devnet` instance, and patch the json exported in step 3 with these new contracts
7. shutdown anvil, writing dump state
8. build a docker image from the `anvil` image with the deploy files and the dumped state.

closes #381, #385

### CODE REVIEW

**Comment by tuler** [comment]
This will also fix #385

File: packages/devnet/contracts/Devnet.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import {IConsensus} from "@cartesi/rollups/contracts/consensus/IConsensus.sol";
+import {Authority} from "@cartesi/rollups/contracts/consensus/authority/Authority.sol";
+import {History} from "@cartesi/rollups/contracts/history/History.sol";
+import {IAuthorityHistoryPairFactory} from "@cartesi/rollups/contracts/consensus/authority/IAuthorityHistoryPairFactory.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IMarketplace} from "@sunodo/contracts/contracts/marketplace/IMarketplace.sol";
+import {IValidatorNodeProvider} from "@sunodo/contracts/contracts/provider/INodeProvider.sol";
+
+contract Devnet {
+    IERC20 public immutable token;
```
Comment (guidanoli) [inline packages/devnet/contracts/Devnet.sol:14]: Is the `token()` function ever called, once the `Devnet` contract is deployed?
If not, we can remove this state variable and the assignment at line 24.

Comment (tuler) [reply packages/devnet/contracts/Devnet.sol]: Removed

**Comment by tuler** [comment]
I'm rethinking if we need this deployment of `DevnetAuthority` and `DevnetHistory` considering the introduction of the `SelfHostedApplicationFactory` of PR https://github.com/sunodo/sunodo/pull/409

We can just call the `SelfHostedApplicationFactory` in the `dapp_deployer` container, and hardcode all addresses (authority, history, dapp) since we can pre-calculate it.

We can use a zero hash as the templateHash, so we have a fixed dapp address, and disable the hash validation of the node using the option CARTESI_FEATURE_DISABLE_MACHINE_HASH_CHECK

**Comment by endersonmaia** [comment]
I suggest this PR should be on top of #308.

**Comment by guidanoli** [comment]
Shouldn't this PR have `feature/adapt-to-rollups-node-1.3.0` as the base branch?

**Comment by tuler** [comment]
I'll close this PR in favor of #409 
With deterministic deployment we don't need the `Devnet` contract deployment, because we can hard-code contract addresses.

**Pull request closed without merging**